sdks/python/apache_beam/dataframe/expressions_test.py

```diff
@@ -53,6 +54,61 @@ def test_expression_proxy_error(self):
     with self.assertRaises(TypeError):
       expressions.ComputedExpression('add', lambda a, b: a + b, [a, b])
 
+  def test_output_partitioning_preserves_singleton(self):
+    a = expressions.PlaceholderExpression(1)
+    b = expressions.PlaceholderExpression(2)
+
+    preserves_singleton = expressions.ComputedExpression(
+        'add',
+        lambda a,
+        b: a + b, [a, b],
+        requires_partition_by=partitionings.Nothing(),
+        preserves_partition_by=partitionings.Singleton())
+
+    for partitioning in (partitionings.Singleton(), ):
+      self.assertEqual(
+          expressions.output_partitioning(preserves_singleton, partitioning),
+          partitioning,
+          f"Should preserve {partitioning}")
+
+    for partitioning in (partitionings.Index([0]),
+                         partitionings.Index(),
+                         partitionings.Nothing()):
+      self.assertEqual(
+          expressions.output_partitioning(preserves_singleton, partitioning),
+          partitionings.Nothing(),
+          f"Should NOT preserve {partitioning}")
+
+  def test_output_partitioning_preserves_index(self):
+    a = expressions.PlaceholderExpression(1)
+    b = expressions.PlaceholderExpression(2)
+
+    preserves_index = expressions.ComputedExpression(
```

Comment:
`preserves_partial_index`? Also, an example would be helpful (e.g. something that replaces level 2 of the index with a column).

Reply:
Done